api/task_manager.cc

```diff
@@ -0,0 +1,165 @@
+                return filter_tasks(task.second, params);
+            });
+            for (auto& [task_id, task] : filtered_tasks) {
+                task_stats ts;
+                ts.task_id = task->id().to_sstring();
+                ts.state = task->get_status().state;
+                local_res.push_back(ts);
+            }
+            return local_res;
+        });
+
+        std::function<future<>(output_stream<char>&&)> f = [r = std::move(res)] (output_stream<char>&& s) -> future<> {
+            bool first = true;
+            auto res = std::move(r);
+            co_await s.write("[");
+            for (auto& v: res) {
+                for (auto& stats: v) {
+                    auto fut = first ? make_ready_future<>() : s.write(", ");
+                    first = false;
+                    co_await std::move(fut);
+                    tm::task_stats ts;
+                    ts = stats;
+                    co_await formatter::write(s, ts);
+                }
+            }
+            co_await s.write("]");
+            co_await s.close();
+        };
+        co_return std::move(f);
+    });
+
+    tm::get_task_status.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        auto task = co_await task_manager::invoke_on_task<task_manager::foreign_task_ptr>(ctx.tm, id, [] (task_manager::task_ptr task) -> future<task_manager::foreign_task_ptr> {
+            auto state = task->get_status().state;
+            if (state == task_manager::task_state::done || state == task_manager::task_state::failed) {
+                task->unregister_task();
+            }
+            co_return std::move(task);
+        });
+        co_return co_await retrieve_status(std::move(task));
+    });
+
+    tm::abort_task.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        co_await task_manager::invoke_on_task(ctx.tm, id, [] (task_manager::task_ptr task) -> future<> {
+            if (!task->is_abortable()) {
+                co_await coroutine::return_exception(std::runtime_error("Requested task cannot be aborted"));
+            }
+            co_await task->abort();
+        });
+        co_return json_void();
+    });
+
+    tm::wait_task.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        task_id id = task_id{utils::UUID{req->param["task_id"]}};
+        auto task = co_await task_manager::invoke_on_task<task_manager::foreign_task_ptr>(ctx.tm, id, [] (task_manager::task_ptr task) {
+            return task->done().then_wrapped([task] (auto f) {
+                task->unregister_task();
+                f.get();
+                return make_ready_future<task_manager::foreign_task_ptr>(make_foreign<task_manager::task_ptr>(task));
```

Comment:
we auto-futurize the results in continuations, so I think
`return make_foreign<task_manager::task_ptr>(task)` should work too, and is simpler.